# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1767820788 -3600
# Parent  e74718d1da56c75d52b523efe69431eefd379068
Feature: Reveal password in mail account password prompt.

diff --git a/mail/base/content/commonDialog.xhtml b/mail/base/content/commonDialog.xhtml
--- a/mail/base/content/commonDialog.xhtml
+++ b/mail/base/content/commonDialog.xhtml
@@ -17,9 +17,9 @@
   <head>
     <title></title>
     <meta
       http-equiv="Content-Security-Policy"
-      content="default-src chrome:; img-src moz-icon: chrome:; style-src chrome: 'unsafe-inline';"
+      content="default-src chrome: 'unsafe-inline'; img-src moz-icon: chrome:; style-src chrome: 'unsafe-inline';"
     />
     <link rel="stylesheet" href="chrome://global/skin/global.css" />
     <link rel="stylesheet" href="chrome://global/content/commonDialog.css" />
     <link rel="stylesheet" href="chrome://global/skin/commonDialog.css" />
@@ -117,9 +117,14 @@
               id="password1Label"
               data-l10n-id="common-dialog-password"
               control="password1Textbox"
             />
+            <xul:hbox>
             <input type="password" id="password1Textbox" dir="ltr" />
+            <button id="password1Toggle" type="button" onclick="passwordToggle(event)">
+              <img id="passwordInfo" src="chrome://messenger/skin/icons/new/compact/hidden.svg"></img>
+            </button>
+            </xul:hbox>
           </div>
           <div id="checkboxContainer" class="dialogRow" hidden="hidden">
             <div />
             <!-- spacer -->
diff --git a/mail/base/content/globalOverlay.js b/mail/base/content/globalOverlay.js
--- a/mail/base/content/globalOverlay.js
+++ b/mail/base/content/globalOverlay.js
@@ -128,4 +128,39 @@ function goSetCommandEnabled(id, enabled
     });
     window.dispatchEvent(commandStateEvent);
   }
 }
+
+// Stuff for toggling the password in commonDialog.xhtml.
+// Copied from mail/components/accountcreation/content/accountSetup.js.
+function passwordToggle(event) {
+  // Prevent the form submission if the user presses Enter.
+  event.preventDefault();
+
+  // The password field is in plain text, change it back to the proper type.
+  if (document.getElementById("password1Textbox").type == "text") {
+    hidePassword();
+  } else {
+    showPassword();
+  }
+}
+
+/**
+ * Convert the password field into a plain text field allowing users and
+ * assistive technologies to read the typed text.
+ */
+function showPassword() {
+  document.getElementById("password1Textbox").type = "text";
+  const toggleImage = document.getElementById("passwordInfo");
+  toggleImage.src = "chrome://messenger/skin/icons/new/compact/eye.svg";
+  toggleImage.classList.add("password-toggled");
+}
+
+/**
+ * Convert the password field back to its default password type.
+ */
+function hidePassword() {
+  document.getElementById("password1Textbox").type = "password";
+  const toggleImage = document.getElementById("passwordInfo");
+  toggleImage.src = "chrome://messenger/skin/icons/new/compact/hidden.svg";
+  toggleImage.classList.remove("password-toggled");
+}
diff --git a/mail/themes/shared/mail/themeableDialog.css b/mail/themes/shared/mail/themeableDialog.css
--- a/mail/themes/shared/mail/themeableDialog.css
+++ b/mail/themes/shared/mail/themeableDialog.css
@@ -559,4 +559,27 @@ hr {
   width: 100%;
   border-top: 1px solid var(--field-border-color);
   border-bottom: 0;
 }
+
+/* Styles for the password dialog. */
+#passwordInfo {
+  -moz-context-properties: fill, fill-opacity, stroke;
+  fill: color-mix(in srgb, currentColor 20%, transparent);
+  stroke: currentColor;
+  fill-opacity: 0.7;
+
+  &.password-toggled {
+    color: var(--button-background-color-primary);
+    fill: color-mix(in srgb, currentColor 20%, transparent);
+    stroke: currentColor;
+  }
+}
+
+button#password1Toggle {
+  padding: 0 3px;
+  margin-left: 0;
+}
+
+input#password1Textbox {
+  width: 100%;
+}
